networking/notes/shelley/CommunicationPatterns.md

```diff
@@ -0,0 +1,283 @@
+
+-- Publish a message. Asynchronous. Returns right away.
+-- The program which publishes is not informed of publication failure.
+-- The pub/sub system can retry or drop according to some policy.
+publish :: PubSub topic d -> Publication topic -> d ()
+
+-- Does nothing if already subscribed and returns False.
+-- The callback is a mechanism for backpressure. The dequeueing thread of the
+-- pub/sub system will wait for the callback to finish before continuing.
+subscribe :: PubSub topic d -> topic k v -> d Bool
+
+-- Does nothing if not currently subscribed.
+unsubscribe :: PubSub topic d -> topic k v -> d ()
+
+-- What to do with a publication.
+-- Announcements must indicate whether to request or ignore.
+-- Payloads must be processed somehow.
+data Next topic d = Next
+  { announce :: NodeId -> topic k v -> k -> d Bool
+  , payload :: NodeId -> topic k v -> v -> d ()
+  }
+
+-- Take the next publication (from any subscriber).
+next :: PubSub topic d -> Next topic d -> d ()
+```
+
+## Questions, caveats, risks
+
+### Publish over subscription channel
+
+It would make a lot of sense to send the published data over the same
+conversation which is used to establish the subscription, rather than (as we
+do now) have one conversation control the lifetime of the connection, and
+starting other conversations to publish each datum.
+
+It's not clear whether this has any big advantages or disadvantages.
+
+Decided: we'll use one time-warp conversation for the whole subscription.
+The subscriber can add/remove topics, ACK published keys, and send keepalive
+messages. The publisher can send keys and values.
+
+### Changes to time-warp
+
+The goal should be to do minimal changes to time-warp. It already enables
+what we need to do, but not in the most efficient or straightforward way.
+
+### Topic filtering
+
+Should this be done at publisher or subscriber?
+Doing it at the publisher will reduce network traffic but the publisher will
+have to do more work. Doing it at the subscriber offloads the filtering work
+to the subscriber. Filtering won't be expensive (a pattern match) so it's
+probably best to do it at the publisher.
+
+Decided: do it publisher side.
+
+### Dead subscriptions due to half-open connections
+
+cardano-sl pull request
+[#1852](https://github.com/input-output-hk/cardano-sl/pull/1852) was made to fix
+[CSL-1676] and is relevant to pub/sub. We'll keep this solution or something
+like it in order to detect dead subscriptions.
+
+### Number of subscribers
+
+Each subscription will be served by one TCP connection, so the OS imposes a
+limit on the number of subscribers. It's plausible that even far below this
+limit performance will start to suffer. But it's also plausible that the number
+of subscriptions actually needed per node is low enough (not more than a
+hundred) to avoid any such problems.
+There's already a maximum subscribers configuration in the outbound queue that
+can be reused.
```

Comment:
@dcoutts said

> We may need something on the inbound side, at the network-transport-tcp level to reject incoming tcp connections, and/or at the time-warp layer to reject after doing the initial peer data handshake. It depends on what basis we need to reject.
> 
> Obviously there's a cost per tcp connection (not just in the app but also the kernel) and rejecting at this stage is cheaper. We will probably need something simple like an absolute limit at the network-transport-tcp level, so that we simply do not accept more connections once we hit a limit. But if we also want to filter on the basis of info then we would need to accept, do the peer data handshake and then close the connection if we are choosing to reject that peer. That would be implemented at the time-warp layer, but we would need a network-transport extension to drop an existing heavyweight connection to a peer.

@avieth said

> https://github.com/haskell-distributed/network-transport/issues/25
> https://github.com/haskell-distributed/network-transport/issues/17